flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentProvider.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.core.memory;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * The provider used for requesting and releasing batch of memory segments.
+ */
+public interface MemorySegmentProvider {
+	/** Requests default number of memory segments. */
+	Collection<MemorySegment> requestMemorySegments() throws IOException;
+
+	Collection<MemorySegment> requestMemorySegments(int numRequiredBuffers) throws IOException;
```

Comment:
If we only consider the semantics of `MemorySegmentProvider` interface, the `requestMemorySegments()` is a bit redundant, which could be replaced by `requestMemorySegments(int numRequiredBuffers)`. The only reason we keep both methods is that it seems not elegant to maintain `networkBuffersPerChannel` in `SingleInputGate`.

TBH, I do not think it is worth defining this new interface not very clean to accommodate the implementation. Because `requestMemorySegments()` seems blurry for the usage which relies on the default value, and we should also consider if the default value is 0, whether the returned collection is null or empty size. 
In addition, `requestMemorySegments(int numRequiredBuffers)` is only used for tests ATM.

Reply:
I think `MemorySegmentProvider` makes sense from point of interface segregation principle. This interface encapsulates segment management concern which is the only thing that remote channel needs atm. `NetworkBufferPool` looks like implementing 2 separate concerns: `BufferPoolFactory` and `MemorySegmentProvider`.

As for `requestMemorySegments(int numRequiredBuffers)`, you are right, it is more clear to remove it from the interface and adjust the tests with explicit default value, e.g. 1.

Reply:
We are indeed in the lack of semantics of requesting/releasing batch of buffers before, and I think it is reasonable to extend these methods. 

- `SegmentProvider` is not really needed because `RemoteInputChannel` still needs construct the `Buffer` internally after requesting `MemorySegment`.

- We could extend the general `BufferProvider` interface to provide method of `requestBuffers(int numRequiredBuffers, BufferRecycler)`,  and `RemoteInputChannel` already implements `BufferRecycler` interface.

- We could also extend the general `BufferRecycler`  interface to provide method of `recycle(Collection<MemorySegment>)`. 

To do so we make use of the existing semantics and architecture. But the only concern is we would affect the previous implementations to bring more changes.